neurolib/models/multimodel/builder/base/network.py

```diff
@@ -0,0 +1,728 @@
+        and its equations. Must be symbolic, i.e. defined with basic math
+        operators and symengine operations on state vector. Should return list
+        as
+        [(se.Symbol, <symbolic definition>)]
+        """
+        raise NotImplementedError
+
+    @property
+    def initial_state(self):
+        """
+        Return initial state of this node, i.e. sum of masses initial states.
+        """
+        return np.array(sum([mass.initial_state for mass in self], [],))
+
+    def all_couplings(self, mass_indices=None):
+        """
+        Return all couplings from masses denoted by index.
+
+        :param mass_indices: indices as to which mass to probe
+        :type mass_indices: list|None
+        :return: coupling from all masses indicated in the mass_indices
+        :rtype: dict
+        """
+        mass_indices = mass_indices or np.arange(len(self.masses)).tolist()
+        all_couplings = {}
+        var_idx = 0
+        for mass_idx, mass in enumerate(self.masses):
+            if mass_idx in mass_indices:
+                all_couplings.update({var_idx + k: v for k, v in mass.coupling_variables.items()})
+            var_idx += mass.num_state_variables
+        return all_couplings
+
+    def _derivatives(self, network_coupling=None):
+        """
+        Gather all derivatives from all masses.
+
+        :param network_coupling: dict of network coupling for this node, if
+            None, default network coupling will be used
+        :type network_coupling: dict|None
+        :return: derivatives of the state vector for this node
+        :rtype: list
+        """
+        if network_coupling is None:
+            network_coupling = self.default_network_coupling
+        node_system_equations = []
+        var_idx = 0
+        for mass in self:
+            # get coupling variables
+            coupling_vars = {
+                self._strip_index(key): value
+                for key, value in self.sync_symbols.items()
+                if self._strip_index(key) in mass.required_couplings
+            }
+            coupling_vars.update(
+                {key: value for key, value in network_coupling.items() if key in mass.required_couplings}
+            )
+            mass.idx_state_var = self.idx_state_var + var_idx
+            node_system_equations += mass._derivatives(coupling_vars)
+            var_idx += mass.num_state_variables
+        assert len(node_system_equations) == self.num_state_variables
+        return node_system_equations
+
+
+class SingleCouplingExcitatoryInhibitoryNode(Node):
+    """
+    Basic node with arbitrary number of excitatory and inhibitory populations,
+    but the coupling is through one variable - usually firing rate of the
+    population. Will compute connectivity within node as per population types.
+    This node definition assumes constant delays, i.e. not dependent on time
+    nor dynamics (state variables).
```

Comment:
In theory, you can have state-dependent delays or time-dependent delays - at least `jitcdde` supports that - that would be tricky in `numba` backend... OK, so when you really want to integrate system like that, you can, but you need to use `jitcdde` backend for it.. should I remove this comment?

Reply:
I would remove it yes, but it's very interesting to know that `jitccde` can handle it. It could be indeed tricky in `numba` but I wouldn't even be surprised if it worked by blindly changing the delay matrix during runtime ... savage!

Reply:
hehe that could work! anyway, removed the comment, now it is explicitly assumed that delays are constant. but we can of course still keep in mind that delays actually, in general, can be state- or time-dependent.